### PR TITLE
WINC-998: Clear proxy variables from Windows nodes

### PR DESCRIPTION
**Pull request by mansikulkarni96**
This PR implements the following scenarios:
1. On existing Windows nodes, when an cluster-wide proxy is removed WICD is able to remove the ENV variables on the instance.
2. When a node deconfiguration occurs, WICD cleanup command will remove the ENV vars set on the instance if cluster-wide proxy is enabled. This includes HTTP_PROXY, HTTPS_PROXY, NO_PROXY being removed from the instance along with the keys set in the Windows registry.

### CODE REVIEW

File: pkg/daemon/cleanup/cleanup.go [hunk 1]
```diff
@@ -65,13 +75,47 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
 		return err
 	}
 	cleanupContainers()
-
+	if len(latestCMData.EnvironmentVars) != 0 {
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:78]: Shouldn't we use the CM tied to `VersionAnnotation`? We could use the latest one as backup like we do in  getServicesToRemove()

File: pkg/daemon/controller/controller.go [hunk 2]
```diff
@@ -353,6 +358,23 @@ func (sc *ServiceController) ensureVarsAreUpToDate(envVars map[string]string) (b
...
+// An instance restart is required after they are removed to ensure all processes pick up the updated values.
+func EnsureEnvVarsAreRemoved(registryKey registry.Key, supportedEnvVars map[string]string) (bool, error) {
+	restartRequired := false
+	for _, envVar := range supportedEnvVars {
+		err := registryKey.DeleteValue(envVar)
+		if err != nil {
+			if err != registry.ErrNotExist {
+				return false, err
+			}
+		} else {
```
Comment (saifshaikh48) [inline pkg/daemon/controller/controller.go:371]: should we also log that the key was successfully removed

File: pkg/daemon/powershell/powershell.go [hunk 3]
```diff
@@ -16,7 +21,7 @@ type commandRunner struct{}
 // Run runs the command with the PowerShell on PATH
 func (r *commandRunner) Run(cmd string) (string, error) {
 	out, err := exec.Command("powershell", "/c", cmd).CombinedOutput()
-	if err != nil {
+	if err != nil && !strings.Contains(err.Error(), cmdExitNoStatus) {
```
Comment (saifshaikh48) [inline pkg/daemon/powershell/powershell.go:24]: Could you explain why this change is needed?

File: pkg/daemon/cleanup/cleanup.go [hunk 4]
```diff
@@ -72,47 +82,99 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
 	if err != nil {
-		return nil, err
+		return err
+	}
+	if len(cmData.EnvironmentVars) == 0 {
+		// no-op
+		return nil
+	}
+	// systemEnvVarRegistryPath is where system level environment variables are stored in the Windows OS
+	const systemEnvVarRegistryPath = `SYSTEM\CurrentControlSet\Control\Session Manager\Environment`
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:106]: same constant exists within a func somewhere in `controller` package. consider making it global over there and referencing that here

File: pkg/daemon/cleanup/cleanup.go [hunk 5]
```diff
@@ -72,47 +82,99 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
+		if closeErr != nil {
+			klog.Errorf("could not close key: %v", closeErr)
+		}
+	}()
+	restartRequired, err := controller.EnsureEnvVarsAreRemoved(registryKey, cmData.SupportedEnvVars)
+	if err != nil {
+		return err
+	}
+	if restartRequired {
+		restartCmd := "Restart-Computer -Force"
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:123]: Please add a comment why restart is necessary

File: pkg/daemon/cleanup/cleanup.go [hunk 6]
```diff
@@ -72,47 +82,99 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
+	if err != nil {
+		return fmt.Errorf("unable to open Windows system registry key %s: %w", systemEnvVarRegistryPath, err)
+	}
+	defer func() { // always close the registry key, without swallowing any error returned before the defer call
+		closeErr := registryKey.Close()
+		if closeErr != nil {
+			klog.Errorf("could not close key: %v", closeErr)
+		}
+	}()
+	restartRequired, err := controller.EnsureEnvVarsAreRemoved(registryKey, cmData.SupportedEnvVars)
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:118]: do you think we should be merging the `SupportedEnvVars` from the versionCM and the latestCM, then pass that list to `EnsureEnvVarsAreRemoved()` (similar to how services are removed)?

File: test/e2e/proxy_test.go [hunk 7]
```diff
@@ -76,6 +78,28 @@ func testEnvVars(t *testing.T) {
...
+	_, err = tc.client.Config.ConfigV1().Proxies().Patch(
+		context.TODO(),
+		"cluster",
+		types.JSONPatchType,
+		patchData,
+		meta.PatchOptions{},
+	)
+	patchString := string(patchData)
+	require.NoErrorf(t, err, "unable to patch %s", patchString)
+	time.Sleep(5 * time.Minute)
```
Comment (saifshaikh48) [inline test/e2e/proxy_test.go:95]: please comment why this sleep is needed. Can we call `waitForConfiguredWindowsNodes()` or something similar instead?

File: pkg/daemon/powershell/powershell.go (hunk 3 above)
Comment (mansikulkarni96) [reply pkg/daemon/powershell/powershell.go:24]: Restart-computer always returns error with cmdExitNoStatus

File: pkg/daemon/cleanup/cleanup.go (hunk 6 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:117]: I looked into it and the reason why we merged services from both the CM's was there were some services which were installed by WMCO and manged by WICD. Since we are going to have all envVars created and cleaned up by WICD from the beginning, do we need to do that?

File: pkg/daemon/cleanup/cleanup.go [hunk 8]
```diff
@@ -57,13 +58,22 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
 		klog.Exitf("could not create service manager: %s", err.Error())
 	}
 	defer svcMgr.Disconnect()
-	services, err := getServicesToRemove(ctx, directClient, node, configMapNamespace)
+	// get data from the latest services ConfigMap
```
Comment (aravindhp) [inline pkg/daemon/cleanup/cleanup.go:61]: You are getting the latest service ConfigMap, passing it to `getServicesToRemove` and `ensureEnvVarsAreRemoved` and within each function getting the version ConfigMap and falling back to the latest if that fails. Shouldn't figuring out which ConfigMap to use be done outside of the two functions? This was we also ensure that both methods are using the same ConfigMap.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @mansikulkarni96. Please address my comments.

File: pkg/daemon/controller/controller.go [hunk 9]
```diff
@@ -321,22 +321,26 @@ func (sc *ServiceController) reconcileEnvironmentVariables(envVars map[string]st
...
-			closeErr := registryKey.Close()
-			if closeErr != nil {
-				klog.Errorf("could not close key: %v", closeErr)
-			}
-		}()
+	}()
 
+	if len(envVars) == 0 {
+		return EnsureEnvVarsAreRemoved(registryKey, supportedEnvVars)
+	}
```
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply pkg/daemon/controller/controller.go]: addressed

File: pkg/servicescm/servicescm.go [hunk 10]
```diff
@@ -27,6 +27,9 @@ const (
 	filesKey = "files"
 	// envVarsKey is an optional key in the services ConfigMap. The value for this key is a string-to-string map.
 	envVarsKey = "environmentVars"
+	// supportedEnvVarsKey is a required key which lists the supported env vars in the services ConfigMap.
```
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply pkg/servicescm/servicescm.go:31]: created https://issues.redhat.com/browse/WINC-1102 as a follow-up.

File: pkg/servicescm/servicescm.go [hunk 11]
```diff
@@ -27,6 +27,9 @@ const (
...
+	// The value for this key is a string-to-string map.
```
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply pkg/servicescm/servicescm.go:32]: I ran into some json masrhalling issues with this approach earlier but I retried it, confirmed it works and updated.

File: pkg/servicescm/servicescm.go [hunk 12]
```diff
@@ -269,6 +290,17 @@ func (cmData *Data) ValidateExpectedContent(expected *Data) error {
 				cmData.EnvironmentVars[expectedEnvVar])
 		}
 	}
+	if len(cmData.SupportedEnvVars) != len(expected.SupportedEnvVars) {
+		return fmt.Errorf("unexpected number of supported environment variable")
+	}
+
+	for i, expectedEnvVar := range expected.SupportedEnvVars {
+		if cmData.SupportedEnvVars[i] != expected.SupportedEnvVars[i] {
```
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply pkg/servicescm/servicescm.go:325]: Right, only way I can think of doing a comparison irrespective of order is doing a deep equal on the slices.

File: pkg/daemon/cleanup/cleanup.go (hunk 8 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:61]: updated, the logic was a bit complex for removing the services, I pulled this out.

File: pkg/daemon/cleanup/cleanup.go (hunk 6 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:119]: @saifshaikh48 WDYT?

File: pkg/daemon/cleanup/cleanup.go [hunk 13]
```diff
@@ -72,47 +82,99 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
+	if err != nil {
+		return err
+	}
+	// rebooting instance to set the environment variables at the process level as expected
+	if restartRequired {
+		restartCmd := "Restart-Computer -Force"
+		cmdRunner := powershell.NewCommandRunner()
+		out, err := cmdRunner.Run(restartCmd)
+		if err != nil {
+			return fmt.Errorf("'%s' failed with out: %s, err: %w", restartCmd, out, err)
```
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go]: This is the only place it's being used here. Other places in the controller where we reconcile and remove vars, we do the reboot and reinitialize through the Windows VM object.

File: pkg/servicescm/servicescm.go [hunk 14]
```diff
@@ -127,6 +135,32 @@ func GetLatest(c client.Client, ctx context.Context, namespace string) (*core.Co
 	return &servicesCMs[0], nil
 }
 
+// GetVersionCM returns the configMap associated with the node's version annotation
```
Comment (aravindhp) [inline pkg/servicescm/servicescm.go:138]: The function is being introduced in this commit but being used in a follow up commit. It should be named `GetVersion()` to avoid stuttering given the call sites will read `servicescm.GetVersion()`.

You don't need to be constrained by the commit prefixes when introducing changes. Introducing them in a logical manner is more important as it makes reviews easier.

File: pkg/servicescm/servicescm.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/servicescm/servicescm.go:325]: This could be prone to false negatives then. Are we okay with that?

File: pkg/daemon/cleanup/cleanup.go (hunk 8 above)
Comment (aravindhp) [reply pkg/daemon/cleanup/cleanup.go:62]: My comment was more of a question rather than a request for a change. In future please respond if it is making the code more complex before making the changes.

File: pkg/daemon/cleanup/cleanup.go [hunk 15]
```diff
@@ -72,47 +96,55 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
 	return nil
 }
 
-// getServicesToRemove returns a list of services that should be removed as part of the cleanup process
-// returns the merged Data of the latest ConfigMap, and the ConfigMap specified by the node's version annotation
-func getServicesToRemove(ctx context.Context, cli client.Client, node *core.Node, configMapNamespace string) ([]servicescm.Service, error) {
-	// get data from the latest services ConfigMap
-	latestCM, err := servicescm.GetLatest(cli, ctx, configMapNamespace)
+// ensureEnvVarsAreRemoved removes all WICD configured ENV variables from this instance
+func ensureEnvVarsAreRemoved(cm *core.ConfigMap) error {
```
Comment (aravindhp) [inline pkg/daemon/cleanup/cleanup.go:100]: Ideally we should be following the same pattern with services i.e. get the env variables to remove and  then remove them. I also don't fully understand the split between `ensureEnvVarsAreRemoved()` and `controller.EnsureEnvVarsAreRemoved() `. To be clear I am not asking for any changes here but just finding the logic split of methods a bit hard to follow across packages.

File: pkg/daemon/cleanup/cleanup.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/daemon/cleanup/cleanup.go]: I find this a bit odd. I will take a look at this deeper outside of this PR.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/servicescm/servicescm.go (hunk 14 above)
Comment (mansikulkarni96) [reply pkg/servicescm/servicescm.go:138]: agreed, apologies I didn't double-check and this was added in a local rebase conflict.

File: pkg/daemon/cleanup/cleanup.go (hunk 15 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:100]: Thanks for the feedback, I ll give it a try if I can make it less complex. I am also working on merging the CM's through a common function, I'll add it to a separate commit for easier review.

Comment (aravindhp) [reply pkg/daemon/cleanup/cleanup.go:100]: I don't think we should add merging of the environment if it's too complex. Plus we don't want to expand the scope of the bug.

File: pkg/servicescm/servicescm.go [hunk 16]
```diff
@@ -95,16 +98,20 @@ type Data struct {
 	Files []FileInfo `json:"files"`
 	// EnvironmentVars will be set as system-level environment variables on each Windows instance
 	EnvironmentVars map[string]string `json:"environmentVars,omitempty"`
+	// SupportedEnvVars contains information about the WMCO supported environment variables
+	SupportedEnvVars []string `json:"supportedEnvVars,omitempty"`
```
Comment (sebsoto) [inline pkg/servicescm/servicescm.go:103]: In general I'm not a huge fan of the `supportedEnvVars` naming, specifically `supported`.
From a technical sense `watchedEnvironmentVariables` would be more fitting.
`support` is more of a business term, while what this api is really doing is directing the WICD controller to ensure the state of specific resources.

**Review by sebsoto — CHANGES_REQUESTED** [review]
Leaving these comments for now, still reviewing the third commit

File: pkg/servicescm/servicescm.go (hunk 14 above)
Comment (sebsoto) [reply pkg/servicescm/servicescm.go:138]: I don't think this function should be added to this package, this package contains general functions for working with the Windows service CMs, IMO the logic introduced in this function is a bit too specialized, due to its use of `node.Annotations[metadata.VersionAnnotation]`. This should 

I don't see it being used anywhere but within cleanup.go, and would fit better there as a helper function.

File: pkg/daemon/cleanup/cleanup.go [hunk 17]
```diff
@@ -57,13 +57,37 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
+		if err != nil {
+			return err
+		}
+		services = latestCMData.Services
+	} else {
+		relevantCM = versionCM
+		services, err = getServicesToRemove(latestCM, versionCM)
+		if err != nil {
+			return err
+		}
```
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:82]: Its not clear why this logic has to be done, which indicates it should be broken out into a separate documented function.

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/daemon/cleanup/cleanup.go (hunk 15 above)
Comment (sebsoto) [reply pkg/daemon/cleanup/cleanup.go:100]: This function doesn't make it clear that it will restart the computer if theres any change.
Please rename this to reflect what it is doing. `removeAllWatchedEnvVarsAndRestart`

File: pkg/daemon/cleanup/cleanup.go [hunk 18]
```diff
@@ -72,47 +96,55 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
-		}
-		versionCMData, err = servicescm.Parse(versionCM.Data)
+	}()
+	restartRequired, err := controller.EnsureEnvVarsAreRemoved(registryKey, cmData.SupportedEnvVars)
+	if err != nil {
+		return err
+	}
+	// rebooting instance to set the environment variables at the process level as expected
+	if restartRequired {
+		restartCmd := "Restart-Computer -Force"
```
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:126]: Should WICD be doing this? Doesn't it signal to WMCO that a restart is necessary in other places?

Comment (aravindhp) [reply pkg/daemon/cleanup/cleanup.go:126]: Sigh, this is why I found this odd. WICD cannot be restarting instances otherwise we could have multiple or all nodes restarting at the same time.

File: pkg/cluster/config.go [hunk 19]
```diff
@@ -357,11 +357,6 @@ func IsProxyEnabled() bool {
 // GetProxyVars returns a map of the proxy variables and values from the WMCO container's environment. The presence of
 // any implies a proxy is enabled, as OLM would have injected them into the operator spec. Returns an empty map otherwise.
 func GetProxyVars() map[string]string {
-	if clusterWideProxyVars != nil {
-		return clusterWideProxyVars
-	}
-	// if clusterWideProxyVars is not already cached, initialize it. We never expect these values to change during
-	// runtime as OLM restarts the operator when the global cluster proxy config changes
```
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply pkg/cluster/config.go:364]: I'll move this to another pre-commit.

File: pkg/servicescm/servicescm.go (hunk 14 above)
Comment (aravindhp) [reply pkg/servicescm/servicescm.go:138]: @sebsoto we are giving conflicting reviews 😄  I asked for it to be moved here given it has to do with the services ConfigMap. I don't see why it can't be added here as a helper. One alternative would be to make the `VersionAnnotation` an input.

File: pkg/daemon/cleanup/cleanup.go (hunk 18 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:126]: AFAIK WICD will not be restating multiple instances through this code path as we only cleanup one instance at a time.

Comment (aravindhp) [reply pkg/daemon/cleanup/cleanup.go:126]: We shouldn't be spreading reboot responsibility in this manner.

File: pkg/servicescm/servicescm.go (hunk 16 above)
Comment (mansikulkarni96) [reply pkg/servicescm/servicescm.go:103]: I don't have strong opinion on this. It made sense to me as WICD supports these set of env Vars. I can update it.

File: pkg/daemon/cleanup/cleanup.go (hunk 18 above)
Comment (sebsoto) [reply pkg/daemon/cleanup/cleanup.go:126]: Or WMCO directly does the restart when it invokes the cleanup command.

Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:126]: The last step of cleanup is deleting node object itself.

Comment (sebsoto) [reply pkg/daemon/cleanup/cleanup.go:126]: In case there is no node to annotate

Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:126]: Are you suggesting WMCO always restart the instance after cleanup is executed and if proxy was enabled?

Comment (aravindhp) [reply pkg/daemon/cleanup/cleanup.go:126]: WMCO is what deletes the node object, so shouldn't it be present? If it is not present, can't we skip rebooting?

File: pkg/daemon/cleanup/cleanup.go (hunk 15 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:100]: I am refactoring this function it will now only return if restart is required and not perform it.

File: pkg/daemon/cleanup/cleanup.go
Comment (aravindhp) [inline pkg/daemon/cleanup/cleanup.go:1]: This comment is regarding the commit message:
>This commit expands the WICD cleanup command to remove
the ENV vars set on the instance if cluster-wide proxy
is enabled in the cluster.

WICD shouldn't care whether cluster-wide proxy is set or not. It should just be handling the watched environment variables irrespective of the proxy settings.

File: pkg/daemon/controller/controller.go [hunk 20]
```diff
@@ -321,8 +321,7 @@ func (sc *ServiceController) reconcileEnvironmentVariables(envVars map[string]st
 
 // ensureVarsAreUpToDate ensures that the proxy environment variables are set as expected on the instance
 // If there's any changes, an instance restart is required to ensure all processes pick up the updated values.
-func (sc *ServiceController) ensureVarsAreUpToDate(envVars map[string]string) (bool, error) {
-	// systemEnvVarRegistryPath is where system level environment variables are stored in the Windows OS
+func ensureVarsAreUpToDate(envVars map[string]string, watchedEnvVars []string) (bool, error) {
 	const systemEnvVarRegistryPath = `SYSTEM\CurrentControlSet\Control\Session Manager\Environment`
```
Comment (aravindhp) [inline pkg/daemon/controller/controller.go]: You should use the `SystemEnvVarRegistryPath` here also.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for reworking this, @mansikulkarni96. 

A general note: I noticed that we are opening a registry key in a function, passing it to another function to actually remove it and then closing it in the calling function. Ideally this should all happen in the same function. But seeing how the code is structured it is going to take bit of a rewrite so lets attempt that later.

File: pkg/daemon/controller/controller.go (hunk 20 above)
Comment (mansikulkarni96) [reply pkg/daemon/controller/controller.go]: this global variable is introduced in  the cleanup commit where it will be used out of this file.

File: test/e2e/delete_test.go [hunk 21]
```diff
@@ -111,6 +115,30 @@ func (tc *testContext) checkNetworksRemoved(address string) (bool, error) {
...
 
+// checkEnvVarsRemoved returns true if the system and service level ENV vars do not exist on the Windows instance
+func (tc *testContext) checkEnvVarsRemoved(address string) (bool, error) {
+	watchedEnvVars := []string{"HTTP_PROXY", "HTTPS_PROXY", "NO_PROXY"}
+	for _, proxyVar := range watchedEnvVars {
+		systemEnvVars, err := tc.getSystemEnvVar(address, proxyVar)
+		if err != nil {
+			return false, fmt.Errorf("error retrieving system level ENV vars: %w", err)
+		}
+		if systemEnvVars[proxyVar] != "" {
```
Comment (saifshaikh48) [inline test/e2e/delete_test.go:126]:
```suggestion
		if _, exists := systemEnvVars[proxyVar]; exists {
```
otherwise this would erroneously pass if the var exists with value ""

File: test/e2e/delete_test.go [hunk 22]
```diff
@@ -111,6 +115,30 @@ func (tc *testContext) checkNetworksRemoved(address string) (bool, error) {
...
+			return false, nil
+		}
+		for _, svcName := range windows.RequiredServices {
+			svcEnvVars, err := tc.getProxyEnvVarsFromService(address, svcName)
+			if err != nil {
+				return false, fmt.Errorf("error retrieving service level ENV vars: %w", err)
+			}
+			if svcEnvVars[proxyVar] != "" {
```
Comment (saifshaikh48) [inline test/e2e/delete_test.go:134]: same as above

File: test/e2e/proxy_test.go [hunk 23]
```diff
@@ -69,13 +68,48 @@ func testEnvVars(t *testing.T) {
...
 					require.NoErrorf(t, err, "error getting environment variables of service %s", svcName)
-					for _, proxyVar := range supportedProxyVars {
+					for _, proxyVar := range watchedEnvVars {
 						assert.Equalf(t, expectedEnvVars[proxyVar], svcEnvVars[proxyVar], "incorrect value for %s", proxyVar)
 					}
 				})
 			}
 		})
 	}
+	var patches []*patch.JSONPatch
```
Comment (saifshaikh48) [inline test/e2e/proxy_test.go:78]: flow looks good but could you add some comments to make it easier to understand? or move some logic into its own testEnvVarRemoval test method?

File: pkg/daemon/cleanup/cleanup.go [hunk 24]
```diff
@@ -163,6 +183,25 @@ func removeServices(svcMgr manager.Manager, services []servicescm.Service) error
 	return nil
 }
 
+// ensureEnvVarsAreRemoved removes all WICD configured ENV variables from this instance
+func ensureEnvVarsAreRemoved(latestCMData, versionCMData *servicescm.Data) (bool, error) {
+	registryKey, err := registry.OpenKey(registry.LOCAL_MACHINE, controller.SystemEnvVarRegistryPath, registry.ALL_ACCESS)
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:188]: is it possible to do this instead to reuse existing code
```
func ensureEnvVarsAreRemoved(latestCMData, versionCMData *servicescm.Data) (bool, error) {
    if versionCMData == nil {
        return controller.EnsureVarsAreUpToDate(map[string]string{}, latestCMData.WatchedEnvironmentVars)
    }
    return controller.EnsureVarsAreUpToDate(map[string]string{}, versionCMData.WatchedEnvironmentVars)
}
```

File: pkg/daemon/cleanup/cleanup.go [hunk 25]
```diff
@@ -57,13 +58,24 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
 	}
-	if err = removeServices(svcMgr, services); err != nil {
+	if err = removeServices(latestCMData, versionCMData, svcMgr); err != nil {
 		return err
 	}
+	restartRequired, err := ensureEnvVarsAreRemoved(latestCMData, versionCMData)
+	if err != nil {
+		return err
+	}
+	// rebooting instance to set the environment variables at the process level as expected
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:72]: set --> unset

File: pkg/daemon/cleanup/cleanup.go [hunk 26]
```diff
@@ -57,13 +58,24 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
...
+	if restartRequired {
+		// Applying the reboot annotation results in an event picked up by WMCO's node controller to reboot the instance
+		if err = metadata.ApplyRebootAnnotation(ctx, directClient, *node); err != nil {
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:75]: Can this lead to a timing issue where
- we apply the reboot annotation onto the node here as part of cleanup
- WICD cleanup returns
- WMCO deletes the node object, before node controller can process reboot annotation event

Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:75]: I have thought about this scenario but haven't run into this issue during my testing. I ll take a look again at this.

File: pkg/daemon/controller/controller.go [hunk 27]
```diff
@@ -357,6 +370,24 @@ func (sc *ServiceController) ensureVarsAreUpToDate(envVars map[string]string) (b
 	return restartRequired, nil
 }
 
+// EnsureEnvVarsAreRemoved ensures that the given environment variables are removed from the instance's Windows registry
+// An instance restart is required after they are removed to ensure all processes pick up the updated values.
+func EnsureEnvVarsAreRemoved(registryKey registry.Key, envVarsToRemove []string) (bool, error) {
+	restartRequired := false
```
Comment (sebsoto) [inline pkg/daemon/controller/controller.go]: Theres too many functions being named similar names throughout the PR, its getting a bit confusing 😅 
I think it would be best to create a separate package `pkg/daemon/envvar`, and move all the functions and constants that deal specifically with interacting with the Windows registry/API to alter environment variables there. This will help to create some structure, and prevent controller.go from becoming too bloated with general helper functions.

File: pkg/daemon/cleanup/cleanup.go [hunk 28]
```diff
@@ -143,7 +161,8 @@ func mergeServices(s1, s2 []servicescm.Service) []servicescm.Service {
 }
 
 // removeServices uses the given manager to remove all the given Windows services from this instance.
-func removeServices(svcMgr manager.Manager, services []servicescm.Service) error {
+func removeServices(latestCMData, versionCMData *servicescm.Data, svcMgr manager.Manager) error {
+	services := getServicesToRemove(latestCMData, versionCMData)
```
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:165]: This is overloading the responsibilities of this function

File: pkg/daemon/cleanup/cleanup.go [hunk 29]
```diff
@@ -72,19 +83,19 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
 	return nil
 }
 
-// getServicesToRemove returns a list of services that should be removed as part of the cleanup process
-// returns the merged Data of the latest ConfigMap, and the ConfigMap specified by the node's version annotation
-func getServicesToRemove(ctx context.Context, cli client.Client, node *core.Node, configMapNamespace string) ([]servicescm.Service, error) {
+// getLatestAndVersionCMData attempts to get the latest and the version CM data that will be used as part of the
+// WICD cleanup process
+func getLatestAndVersionCMData(ctx context.Context, cli client.Client,
+	namespace string, node *core.Node) (*servicescm.Data, *servicescm.Data, error) {
```
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:89]: I'm having a hard time understand why this change is being made. Is it because we no longer only need to merge the service list from each CM, but we also need to know the watched environment variables?
If that is the case instead of making this change, I suggest you change the existing function to return a merged servicescm.Data, instead of two separate CM. Because that is what is actually desired by the caller of this function. Doing this will allow you simplify the `removeServices()` and `cleanup.ensureEnvVarsAreRemoved()` functions as well

File: pkg/daemon/controller/controller.go (hunk 27 above)
Comment (mansikulkarni96) [reply pkg/daemon/controller/controller.go]: @sebsoto I have no issues with creating another package. However, I want to understand the confusing part here. This is the only function introduced in controller package in this PR.  All the other functions related to env vars are pre-existing only they have been now exported or their params have changed through this PR.
The only other ensureEnvVarsAreRemoved exists in cleanup package as that basically goes through the same logic flow calling the functions in this package.

File: pkg/daemon/cleanup/cleanup.go (hunk 29 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:89]: @sebsoto this has to be done because we need the latest and version CM to be retrieved for both service deletion and environment var deletion.  I would have done what you are suggesting in first place but we only need the merged CM data for services removal and not for env Var Removal. In the previous discussions on this PR, we have decided not to use merged configmap data for env var removal as it is not a complete solution. Here, if the vesion CM data is not available we fall-back to using just the latest CM data.

File: pkg/daemon/cleanup/cleanup.go (hunk 28 above)
Comment (mansikulkarni96) [reply pkg/daemon/cleanup/cleanup.go:165]: It was in deconfigure where it was only used to get the services to be removed. I could move it back but I think it is better suited where it is used instead of passing it in.

File: pkg/daemon/controller/controller.go (hunk 27 above)
Comment (sebsoto) [reply pkg/daemon/controller/controller.go]: The confusing part is across the different commits in this PR, there are 5 separate changes to the function signatures of `controller.ensureEnvVarsAreRemoved`, `ensureVarsAreUpToDate` , and `cleanup.ensureEnvVarsAreRemoved`. Some of that is churn and its hard to keep track of when reviewing.

File: pkg/daemon/cleanup/cleanup.go (hunk 29 above)
Comment (sebsoto) [reply pkg/daemon/cleanup/cleanup.go:90]: > we need the latest and version CM to be retrieved for both service deletion and environment var deletion

Yes, I see that both of them are passed into both `removeServices` and `ensureEnvVarsAreRemoved`

> but we only need the merged CM data for services removal and not for env Var Removal... we have decided not to use merged configmap data for env var removal as it is not a complete solution

Looking at the code within this PR its not clear to me why that is the case. I don't see how services and envVars would differ in the way they should be handled here. Wouldn't it be safer to use a merged list for the same reasons we need to for services? Partially configured instances which are upgraded will potentially have envVars set by previous versions of WMCO which we need to undo.

File: pkg/daemon/controller/controller.go (hunk 27 above)
Comment (mansikulkarni96) [reply pkg/daemon/controller/controller.go]: I have moved envVar related functions into separate package. Please lmk if that helps.

File: pkg/daemon/cleanup/cleanup.go [hunk 30]
```diff
@@ -142,8 +147,20 @@ func mergeServices(s1, s2 []servicescm.Service) []servicescm.Service {
...
+func mergeCMData(data1, data2 map[string]string) map[string]string {
+	mergedData := make(map[string]string)
+	for key, value := range data1 {
+		mergedData[key] = value
+	}
+	for key, value := range data2 {
+		mergedData[key] = value
+	}
+	return mergedData
+}
```
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:160]: Don't you need to merge the services and envvars independently, and then create a new CM from the merged values? Doing this seems like it will just take everything (files, services, etc) from mergedData, as the keys will be the same.

File: pkg/daemon/cleanup/cleanup.go [hunk 31]
```diff
@@ -142,8 +147,20 @@ func mergeServices(s1, s2 []servicescm.Service) []servicescm.Service {
...
+
 // removeServices uses the given manager to remove all the given Windows services from this instance.
-func removeServices(svcMgr manager.Manager, services []servicescm.Service) error {
+func removeServices(services []servicescm.Service, svcMgr manager.Manager) error {
```
Comment (sebsoto) [inline pkg/daemon/cleanup/cleanup.go:163]: can revert this change

File: pkg/daemon/cleanup/cleanup.go (hunk 26 above)
Comment (saifshaikh48) [reply pkg/daemon/cleanup/cleanup.go:75]: Thanks, concern has been resolved. This was addressed by waiting for reboot annotation removal before deleting node object

**Review by sebsoto — APPROVED** [review]
Thanks for the hard work. LGTM

File: pkg/daemon/cleanup/cleanup.go [hunk 32]
```diff
@@ -72,13 +84,15 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
 	return nil
 }
 
-// getServicesToRemove returns a list of services that should be removed as part of the cleanup process
-// returns the merged Data of the latest ConfigMap, and the ConfigMap specified by the node's version annotation
-func getServicesToRemove(ctx context.Context, cli client.Client, node *core.Node, configMapNamespace string) ([]servicescm.Service, error) {
+// getmergedCMData attempts to get the latest and the version CM data specified by the node's version annotation
+// It returns the merged CM Data containing services and the watched environment variables
+func getmergedCMData(ctx context.Context, cli client.Client,
```
Comment (saifshaikh48) [inline pkg/daemon/cleanup/cleanup.go:89]: nit: `getmergedCMData` --> `getMergedCMData`

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

**Comment by mansikulkarni96** [comment]
/retest-required
cluster spinup failures

**Comment by mansikulkarni96** [comment]
/retest-required
I see a [green run](https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-openshift-windows-machine-config-operator-master-vsphere-proxy-e2e-operator) on the job board for another PR
